sqlserver/datadog_checks/sqlserver/stored_procedures.py

```diff
@@ -0,0 +1,186 @@
+        )
+        self._state = StatementMetrics()
+        self._conn_key_prefix = "dbm-procedures"
+        self._procedure_metrics_query = None
+        self._last_stats_query_time = None
+        self._max_procedure_metrics = check.procedure_metrics_config.get("max_procedures", 250)
+
+    def _close_db_conn(self):
+        pass
+
+    def _get_available_procedure_metrics_columns(self, cursor, all_expected_columns):
+        cursor.execute("select top 0 * from sys.dm_exec_procedure_stats")
+        all_columns = {i[0] for i in cursor.description}
+        available_columns = [c for c in all_expected_columns if c in all_columns]
+        missing_columns = set(all_expected_columns) - set(available_columns)
+        if missing_columns:
+            self.log.debug(
+                "missing the following expected query metrics columns from dm_exec_procedure_stats: %s", missing_columns
+            )
+        self.log.debug("found available sys.dm_exec_procedure_stats columns: %s", available_columns)
+        return available_columns
+
+    def _get_procedure_metrics_query_cached(self, cursor):
+        if self._procedure_metrics_query:
+            return self._procedure_metrics_query
+        available_columns = self._get_available_procedure_metrics_columns(cursor, SQL_SERVER_PROCEDURE_METRICS_COLUMNS)
+        self._procedure_metrics_query = PROCEDURE_METRICS_QUERY.format(
+            procedure_metrics_columns=', '.join(['sum({}) as {}'.format(c, c) for c in available_columns]),
+            limit=self.dm_exec_procedure_stats_row_limit,
+        )
+        return self._procedure_metrics_query
+
+    @tracked_method(agent_check_getter=agent_check_getter, track_result_length=True)
+    def _load_raw_procedure_metrics_rows(self, cursor):
+        self.log.debug("collecting sql server procedure metrics")
+        procedure_metrics_query = self._get_procedure_metrics_query_cached(cursor)
+        now = time.time()
+        self._last_stats_query_time = now
+        self.log.debug("Running query [%s]", procedure_metrics_query)
+        cursor.execute(procedure_metrics_query)
+        columns = [i[0] for i in cursor.description]
+        # construct row dicts manually as there's no DictCursor for pyodbc
+        rows = [dict(zip(columns, row)) for row in cursor.fetchall()]
+        return rows
+
+    def _collect_metrics_rows(self, cursor):
+        rows = self._load_raw_procedure_metrics_rows(cursor)
+        rows = [r for r in rows if r.get('procedure_name') and r.get('database_name')]
+        if not rows:
+            return []
+        metric_columns = [c for c in rows[0].keys() if c.startswith("total_") or c == 'execution_count']
+        rows = self._state.compute_derivative_rows(rows, metric_columns, key=_row_key)
+        return rows
+
+    def _to_metrics_payload(self, rows, max_queries):
+        # sort by total_elapsed_time and return the top max_queries
+        rows = sorted(rows, key=lambda i: i['total_elapsed_time'], reverse=True)
```

Comment:
Is there a reason we're applying limits here instead of in the backend like we do for query metrics?

Reply:
🤷  We also do this in the agent for query metrics.

See here https://github.com/DataDog/integrations-core/blob/master/sqlserver/datadog_checks/sqlserver/statements.py#L358

Reply:
Oh yeah, look at that. I guess this is fine then.